File: src/concurrency/book2/sample/chapter8/java/DenyPolicy.java
package concurrency.book2.sample.chapter8.java;

@FunctionalInterface
public interface DenyPolicy {

    void reject(Runnable runnable, ThreadPool threadPool);

    //直接丢弃任务
    class DiscardDenyPolicy implements DenyPolicy {
        @Override
        public void reject(Runnable runnable, ThreadPool threadPool) {
            //do nothing
        }
    }

    //向任务提交者抛出异常
    class AbortDenyPolicy implements DenyPolicy {

        @Override
        public void reject(Runnable runnable, ThreadPool threadPool) {
            throw new RunnableDenyException("The runnable " + runnable + " will be abort.");
        }
    }

    //任务在提交者所在的线程中执行
    class RunnerDenyPolicy implements DenyPolicy {

        @Override
        public void reject(Runnable runnable, ThreadPool threadPool) {
            if (!threadPool.isShutDown()) {
                runnable.run();
            }
        }
    }


}

File: src/concurrency/book2/sample/chapter8/kotlin/DenyPolicy.kt
//package concurrency.book2.sample.chapter8.kotlin
//
//interface DenyPolicy {
//
//    fun reject(runnable: Runnable, pool: ThreadPool)
//
//    class DiscardDenyPolicy:DenyPolicy{
//        override fun reject(runnable: Runnable, pool: ThreadPool) {
//            // do nothing
//        }
//
//    }
//
//    class AbortDenyPolicy:DenyPolicy{
//        override fun reject(runnable: Runnable, pool: ThreadPool) {
//        }
//    }
//
//    class RunnerDenyPolicy:DenyPolicy{
//        override fun reject(runnable: Runnable, pool: ThreadPool) {
//
//        }
//
//    }
//}